scipy/sparse/_base.py

```diff
@@ -475,18 +489,85 @@ def asformat(self, format, copy=False):
     ####################################################################
 
     def multiply(self, other):
-        """Point-wise multiplication by another array/matrix."""
+        """Element-wise multiplication by another array/matrix."""
         if isscalarlike(other):
             return self._mul_scalar(other)
-        return self.tocsr().multiply(other)
+
+        if not (issparse(other) or isdense(other)):
+            # If it's a list or whatever, treat it like an array
+            other_a = np.asanyarray(other)
+            if other_a.ndim == 0 and other_a.dtype == np.object_:
+                return NotImplemented
+            try:
+                other.shape
+            except AttributeError:
+                other = other_a
+
+        if self.ndim < 3 and other.ndim < 3:
+            return self.tocsr().multiply(other)
```

Comment:
non-blocking: Ideally we should have this base class provide all the public API methods, and these owuld dispatch exclusively to private methods (similar to how we already handle `_mul_scalar`). That way, each subclass can have stronger guarantees about the types and shapes of its inputs.

Reply:
I have moved the public `multiply` from `_compressed.py` to `_base.py` as described.

The `isscalarlike` case is now handled in `_base.py` before dispatching to csr's `_multiply_2d_with_broadcasting`. 

As we improve broadcasting support, we can/should move more out of `csr` and into `base`. For example, the case `isdense(other)` doesn't need to convert to CSR -- once it is unwrapped from all the broadcasting code we can call it from `_base`.

In summary, we now have all the public API methods we've been messing with in _base.py and they dispatch to private methods in the subclasses.  A followup PR can attempt to do this with other methods not considered here (like `sum` and `diagonal` and maybe others).